File: server/services/matcher.ts
import { Job, MatchResult, ParsedCV } from "@shared/schema";

export function matchProfileToJobs(profile: ParsedCV, jobs: Job[]): MatchResult[] {
  return jobs.map(job => {
    // 1. Calculate Skill Match
    const jobSkills = job.skillsRequired.map(s => s.toLowerCase());
    const userSkills = new Set(profile.skills.map(s => s.toLowerCase()));
    
    const matchedSkills = jobSkills.filter(skill => userSkills.has(skill));
    const missingSkills = jobSkills.filter(skill => !userSkills.has(skill));
    
    // Weighted score: 
    // Skills count for 70%
    // Experience match counts for 30% (simple check)
    
    const skillScore = jobSkills.length > 0 ? (matchedSkills.length / jobSkills.length) : 0;
    
    // Experience check (Job doesn't explicitly have years in DB schema structured, but we can infer or ignore for MVP)
    // Let's assume Senior roles need 5+, others 2+
    const isSenior = job.title.toLowerCase().includes("senior");
    const reqYears = isSenior ? 5 : 2;
    const expScore = profile.yearsExperience >= reqYears ? 1 : (profile.yearsExperience / reqYears);
    
    let totalScore = (skillScore * 0.7) + (expScore * 0.3);
    
    // Boost if title matches
    const titleMatch = profile.jobTitles.some(t => job.title.toLowerCase().includes(t));
    if (titleMatch) totalScore += 0.1;

    // Cap at 100% (or 1.0)
    totalScore = Math.min(totalScore, 1);
    
    // Generate Insights
    const strengthAlignment = [];
    if (matchedSkills.length > 3) strengthAlignment.push("Strong technical stack overlap");
    if (titleMatch) strengthAlignment.push("Directly relevant job history");
    if (expScore >= 1) strengthAlignment.push("Meets experience requirements");
    
    const improvements = [];
    if (missingSkills.length > 0) {
      improvements.push(`Consider learning or highlighting: ${missingSkills.slice(0, 3).join(", ")}`);
    }
    if (expScore < 1) {
      improvements.push("This role might require more seniority than detected.");
    }

    return {
      jobId: job.id,
      matchPercentage: Math.round(totalScore * 100),
      missingSkills,
      matchedSkills,
      strengthAlignment,
      improvements
    };
  }).sort((a, b) => b.matchPercentage - a.matchPercentage);
}

File: backend/services/matcher.py
from __future__ import annotations

from typing import Iterable, List, Optional

from schemas import MatchJobSummary, MatchResult, ParsedCV


def normalize_skill(skill: str) -> str:
    return skill.strip().lower()


def title_matches(profile_titles: Iterable[str], job_title: str) -> bool:
    job_title_lower = job_title.lower()
    return any(title.lower() in job_title_lower for title in profile_titles)


def _matches_any(needle_text: str, prefs: Iterable[str]) -> bool:
    text = (needle_text or "").lower()
    return any(p and p.lower() in text for p in prefs)


def match_profile_to_jobs(
    profile: ParsedCV,
    jobs: list[dict],
    preferred_locations: Optional[List[str]] = None,
    preferred_arrangements: Optional[List[str]] = None,
    preferred_employment: Optional[List[str]] = None,
) -> List[MatchResult]:
    pref_locs = preferred_locations or []
    pref_arrs = preferred_arrangements or []
    pref_emps = preferred_employment or []

    results: List[MatchResult] = []
    for job in jobs:
        title = job.get("title", "")
        skills_required = job.get("skills_required") or []

        normalized_req = {normalize_skill(s) for s in skills_required}
        normalized_profile_skills = {normalize_skill(s) for s in profile.skills}

        matched_skills = sorted(normalized_req.intersection(normalized_profile_skills))
        missing_skills = sorted(normalized_req.difference(normalized_profile_skills))

        if normalized_req:
            skill_score = len(matched_skills) / len(normalized_req)
        else:
            # Neutral score when no required skills are provided
            skill_score = 0.5

        req_years = 5 if "senior" in title.lower() else 2
        exp_score = min(profile.years_experience / req_years if req_years else 0, 1.0)

        title_bonus = 0.1 if title_matches(profile.job_titles, title) else 0.0

        # Preference bonuses — small additive weights so a perfect-skill match
        # still wins even if it's in the "wrong" city, but two equal matches
        # break ties by location/arrangement/type alignment.
        job_location = job.get("location", "") or ""
        job_arrangement = (job.get("arrangement") or job.get("location") or "")
        job_type = job.get("type", "") or ""

        location_bonus = 0.05 if pref_locs and _matches_any(job_location, pref_locs) else 0.0
        arrangement_bonus = (
            0.05 if pref_arrs and _matches_any(job_arrangement + " " + job_type, pref_arrs) else 0.0
        )
        employment_bonus = 0.03 if pref_emps and _matches_any(job_type, pref_emps) else 0.0

        total_score = min(
            (skill_score * 0.7)
            + (exp_score * 0.3)
            + title_bonus
            + location_bonus
            + arrangement_bonus
            + employment_bonus,
            1.0,
        )
        match_percentage = round(total_score * 100)

        if match_percentage < 10:
            continue

        strength_alignment: list[str] = []
        if len(matched_skills) > 3:
            strength_alignment.append("Strong technical stack overlap")
        if title_bonus > 0:
            strength_alignment.append("Directly relevant job history")
        if exp_score >= 1:
            strength_alignment.append("Meets experience requirements")
        if location_bonus > 0:
            strength_alignment.append(f"Matches your preferred location ({job_location})")
        if arrangement_bonus > 0:
            strength_alignment.append("Matches your preferred work arrangement")
        if employment_bonus > 0:
            strength_alignment.append("Matches your preferred employment type")

        improvements: list[str] = []
        if missing_skills:
            improvements.append(f"Consider learning or highlighting: {', '.join(missing_skills[:3])}")
        if exp_score < 1:
            improvements.append("This role might require more seniority than detected.")
        if pref_locs and location_bonus == 0:
            improvements.append(
                f"Outside your preferred locations ({', '.join(pref_locs[:2])})."
            )

        summary = MatchJobSummary(
            title=title,
            company=job.get("company", ""),
            location=job.get("location", ""),
            is_live=job.get("is_live", False),
            source_name=job.get("source_name"),
        )

        results.append(
            MatchResult(
                job_id=job.get("id", 0),
                match_percentage=match_percentage,
                matched_skills=matched_skills,
                missing_skills=missing_skills,
                strength_alignment=strength_alignment,
                improvements=improvements,
                job=summary,
            )
        )

    results.sort(key=lambda r: r.match_percentage, reverse=True)
    return results
